Full rework and new chapters

bertrand-marquis opened the pull request:
This is a full rework of the specification with the help of chatgpt for the wording and sentences to:
- have chapters more compliants with other transport
- switch to the MUST/SHALL/SHOULD terminology
- clear up some concepts

This is a work in progress but looking more close to a spec.
Any comments are welcome as always :-)

bertrand-marquis committed:
Introduce a protocol version and a message for it

Introduce the VIRTIO_MSG_VERSION and define a protocol to negociate a
protocol version between the driver and the device side.

Current version is set to 0.1.

Bump all message number so that the Version message is numbered 0x2 (0x1
being the Error message).

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
Make the messge size variable

Modify the message protocol to have a variable message size and allow
implementations that could tranfer bigger messages to benefit of all the
size available:
- add a maximum message size in the version message: to inform both
  sides of what the other can support and limit the size used to what
  can be transfered.
- add a message payload size field in the header
- modify get/set feature and config to have a variable size and be able
  to transfer as much data as possible (rework the offset to support any
  offset value and adapt to the different possible sizes)
- rework bus device discovery message to be able to use a bigger bitmap

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
Introduce the configuration generation count

Introduce a configuration generation count and describe it in a specific
chapter.

Modify the following messages to use it:
- get_config: provide the value in the answer
- set_config: add in request, provide in answer, error if not right in
  request.
- get_config_gen: remove message
- event_config: provide the value

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
Give hints in bus message of what can be done

Mention that a bus implementation can provide a solution for memory
sharing or could rely on MSI-x for notification instead of actually
sending an event.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
Rework version message

Change the version from major.minor to an unsigned integer that we will
increase each time the protocol is modified and define current version
as "1".
Remove the version for the device info message as we have a specific
message for it.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
Update the message size definition

Update the message size definition to be clearer:
- minimum message size is 8 bytes (header and empty payload)
- minimum message size that a transport must support is 44 bytes
  (virtqueue set message)

Add a sentence to explain how size negocation works.

Modify get/set feature to use 32 bit blocks instead of allowing to
request any bit offest and size.

Modify bus message to discover devices to also include a number of
device IDs requested in the message (instead of relying on the message
size) to be more coherent with the other messages.

Add a mention in what a bus can do for restricting size and protocol
version that are negociated to restrict them to what it can handle.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
Add a mention of the event message for the config cache

Add a sentence to signal that any change of the configuration not
requested by the driver is signaled to the driver by the device through
the event message.

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

bertrand-marquis committed:
Major rework of the transport-msg spec

Rework the spec to comply more to other transport definitions:
- create an introduction chapter
- create a basic concept chapter
- create bus operation and device initialization and operation chapters
- rework transport and bus messages
- attempt at a "compliance chapter"
- reuse spec macros and tries to switch to may/must/shall/should wording

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

ahomescu commented on transport-msg.tex:
```diff
@@ -1,55 +1,246 @@
+        devices. It can detect available devices through:
+    \begin{itemize}
+      \item Optional message-based queries (\busref{GET_DEVICES}),
+      \item External data sources (e.g., device tree, ACPI tables, hypervisor
```
Is this where `FFA_PARTITION_INFO_GET` would fall?

bertrand-marquis replied:
When using the FF-A virtio-msg bus the idea is that:
- you use FFA_PARTITION_INFO_GET on the FF-A virtio-msg bus frontend with Virtio Backend UUID to find the endpoints on the system that are hosting a virtio-msg bus backend.
- you then use VIRTIO_MSG_GET_DEVICES to retrieve the list of virtio devices available on an endpoint

arnopo commented on transport-msg.tex:
```diff
@@ -276,23 +294,26 @@ \subsection{Messages Definition}\label{sec:Virtio Transport Options / Virtio Ove
+on the maximum message size supported by both sides.
+
+The device answers with the actual configuration value at the offset after the
+modification to give a chance to the driver a chance to detect changes that
```
to give a chance to the driver to detect changes

bertrand-marquis replied:
ack

arnopo commented on transport-msg.tex:
```diff
@@ -453,6 +469,9 @@ \subsection{Messages Definition}\label{sec:Virtio Transport Options / Virtio Ove
 the configuration and discover what has changed.
 
+The maximum size of data that can be communicated with an event depends on the
+maximum message size supported by both sides.
```
suggestion only
```
The amount of data that can be communicated with an event depends on the
maximum message size supported by both sides.
```

bertrand-marquis replied:
ack

arnopo commented on transport-msg.tex:
```diff
@@ -633,11 +652,10 @@ \subsubsection{Message Definition}\label{sec:Virtio Transport Options / Virtio O
 Answer & 0 & 2 & Offset \\
        & 2 & 2 & Next offset or 0 if no devices after Offset \\
-       & 4 & 32 & Bit[n]: Device[offset*256 + n] not available(0)/available(1) \\
+       & 4 & .. & Bit[n]: Device[offset + n] not available(0)/available(1) \\
```
Even if  we have only one device ID, the driver has to parse all the bits. Should we add a `last_offset` field to optimize the parsing?

bertrand-marquis replied:
This is a once per device operation and most of the time there will not be so much entries per message and it could be that there is one at the beginning and one at the end with nothing in the middle. I think that adding a last_offset might be a bit to much here.

arnopo commented on transport-msg.tex:
```diff
@@ -324,31 +350,13 @@ \subsection{Messages Definition}\label{sec:Virtio Transport Options / Virtio Ove
-Request & 0 & 0 & None \\
-\hline
-Answer & 0 & 4 & Atomicity value \\
+& 8 & 4 & Configuration generation count \\
```
Needed if we return an error if the Configuration generation count  sent by the driver is not aligned?

bertrand-marquis replied:
It is needed on the first read of the configuration to have an initial value and also in the case where a device would not generate events. The data contain in the message is only coherent for the generation count contained in the message so we definitely need to have it here.

arnopo commented on transport-msg.tex:
```diff
@@ -1,59 +1,246 @@
+    \end{itemize}
+    Once a device is identified—regardless of discovery method—the bus typically
+    uses \msgref{GET_DEVICE_INFO} to read its device ID and vendor ID, then
+    \emph{registers} that device with the host OS so the usual virtio driver
```
I would remove  `OS` , can be baremetal

bertrand-marquis replied:
definitely a place where we should just say "frontend" without any more qualification.

arnopo commented on transport-msg.tex:
```diff
@@ -1,59 +1,246 @@
+        available devices. Once a device is recognized, the driver uses the
+        common transport messages to perform feature negotiation, configuration,
+        and virtqueue setup.
+  \item \textbf{Device Backends}: Implement virtio device functionality and
```
```suggestion
\item \textbf{Device (Backends)}: Implement virtio device functionality and
```

bertrand-marquis replied:
ack

arnopo commented on transport-msg.tex:
```diff
@@ -1,59 +1,246 @@
+virtio-msg operates around two layers:
+
+\begin{enumerate}
+  \item \textbf{Bus Layer}: A \emph{bus instance} exposes zero or more virtio
```
Does a bus that expose 0 devices make sense? I would expect here `one or several virtio device`

wmamills replied:
A bus could start with zero devices and then add devices via hotplug.  I think the wording is OK as is.

arnopo replied:
If the main information here is that a bus can exist without device, then  the wording is good, 
However, if the information is that a bus supports more than one virtio device, then the "zero devices" is quite disturbing. In such case I would prefer `A \emph{bus instance} may expose several virtio devices`.

That said it is nitpicking. keeping current wording is fine for me

bertrand-marquis replied:
A bus could start with zero devices or end up at some point with 0 devices.
At least in the Vm to VM case, devices could be created dynamically after the VM is created.
I agree with the "may expose" wording, it looks nicer.

arnopo commented on transport-msg.tex:
```diff
@@ -1,59 +1,246 @@
+        different capabilities (e.g., transport revision, max message size) and
+        may discover devices via a message-based or external mechanism. Every
+        discovered device is assigned a \emph{device ID}.
+  \item \textbf{Driver (Frontend)}: Communicates with the bus to learn about
```
As discussed during meeting need to change to `bus device side` and clarify the topology

bertrand-marquis replied:
as said during the meeting we need a global rewording on bus device side or backend side and make it coherent in the complete document.
From my notes we agreed to use frontend and backend.

bertrand-marquis committed:
Minor fixes on the improve version

- set feature bit blocks to 32bit (and not 256 bits)
- readd the optional feature bits in revision
- remove max_size and replace with proper naming
- minor typos and fixes

Signed-off-by: Bertrand Marquis <bertrand.marquis@arm.com>

arnopo reviewed (commented):
As suspected, I forgot to post my comments :smiley:

wmamills commented:
As discussed in the meeting on April 10th, this PR will be accepted as is and and fixes needed will be done on top.
